Cypress tests replace cy.wait with cy.intercept or dynamic waits

Rishith25 opened the pull request:
## Proposed Changes

Fixes #8925 
- Replaced hardcoded cy.wait() with cy.intercept() for API requests
- Implemented dynamic waits to reduce flakiness in tests
- Improved overall test reliability

@ohcnetwork/care-fe-code-reviewers

## Merge Checklist

- [ ] Add specs that demonstrate bug / test a new feature.
- [ ] Update [product documentation](https://docs.ohc.network).
- [ ] Ensure that UI text is kept in I18n files.
- [ ] Request for Peer Reviews
- [ ] Completion of QA


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
	- Enhanced test coverage for patient consultations, discharges, and log updates.
	- Added new methods for intercepting and verifying API requests across various functionalities.
	- Introduced dynamic waits and visibility checks to improve test reliability.

- **Bug Fixes**
	- Corrected success notification messages for inventory management.
	- Improved error handling for invalid inputs during user management.

- **Chores**
	- Removed static wait commands in favor of intercepts for better performance and responsiveness.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

Rishith25 committed:
Cypress Tests cy.wait() replace with cy.intercept() or dynamic wait

Rishith25 committed:
Refactor Cypress tests: replaced static cy.wait() with dynamic waits and removed irrelevant waits

Jacobjeevan commented:
Have you committed all the changes? 🤔  I'm not seeing any added cy.intercept calls in the changeset.

Rishith25 commented:
@Jacobjeevan I reviewed all the Cypress test cases, and it looks like the necessary cy.intercept() calls were already in place. I went ahead and removed the cy.wait() calls where they were unnecessary and added two dynamic waits in specific files to ensure smoother test execution. If I misunderstood anything or missed any cy.intercept() implementations, please let me know.

nihal467 commented on cypress/e2e/assets_spec/AssetHomepage.cy.ts:
```diff
@@ -100,7 +100,6 @@ describe("Asset Tab", () => {
 
   it("Export asset", () => {
     assetPage.selectassetimportbutton();
-    cy.wait(2000);
```
the intercept and API Verification is missing here

nihal467 commented on cypress/e2e/facility_spec/FacilityInventory.cy.ts:
```diff
@@ -43,7 +45,6 @@ describe("Inventory Management Section", () => {
     // verify the last entry deletion
     facilityPage.verifyStockInRow("#row-0", "Added Stock");
     facilityPage.verifyStockInRow("#row-1", "Used Stock");
-    cy.wait(3000);
```
the intercept and API Verification is missing here

nihal467 commented on cypress/e2e/facility_spec/FacilityInventory.cy.ts:
```diff
@@ -59,7 +60,6 @@ describe("Inventory Management Section", () => {
     facilityPage.verifyBadgeWithText(".badge-danger", "Low Stock");
     // modify with manual minimum badge
     facilityPage.clickAddMinimumQuanitity();
-    cy.wait(3000);
```
the intercept and API Verification is missing here

nihal467 commented on .env:
```diff
@@ -8,6 +8,7 @@ REACT_PUBLIC_URL=https://care.ohc.network
 
 # Care API URL without the /api prefix
 REACT_CARE_API_URL=https://careapi.ohc.network
+# REACT_CARE_API_URL=http://localhost:9000
```
revert this change

nihal467 reviewed (changes requested):
- According to the issue, you should replace cy.wait with cy.intercept() to verify that the relevant APIs are being fetched. However, in your PR, you only removed cy.wait without adding any cy.intercept to the code. Please make the necessary changes.
- Make sure that post your changes, cypress action also passes fully

nihal467 commented:
@Rishith25 The cy.wait command was initially added in the test to prevent flakiness. The issue was created to replace hardcorded wait with cy.intercept, ensuring that tests dynamically wait for the necessary API to load and verify that the page is fully loaded before executing the next step in the test.

Rishith25 committed:
Merge branch 'develop' into cypress-tests-replace-cy.wait-with-cy.intercept

nihal467 committed:
Merge branch 'develop' into cypress-tests-replace-cy.wait-with-cy.intercept

nihal467 commented:
@Rishith25 what is the status on this PR

Rishith25 commented:
@nihal467  I am having exams I will try to work from Saturday and update the PR

Rishith25 committed:
Resolved merge conflicts

Rishith25 committed:
Cypress test cases

nihal467 commented:
@Rishith25 what's the update

Rishith25 commented:
I am working on it

Rishith25 committed:
Added intercept in Facility Inventory

Rishith25 committed:
Merge branch 'cypress-tests-replace-cy.wait-with-cy.intercept' of https://github.com/Rishith25/care_fe into cypress-tests-replace-cy.wait-with-cy.intercept

Rishith25 committed:
reverting .env file

nihal467 commented:
@Rishith25 what is the update on this PR

Rishith25 committed:
Merge branch 'develop' into cypress-tests-replace-cy.wait-with-cy.intercept

Rishith25 committed:
cy.wait is replaced with cy.intercept in ManageUser

Rishith25 committed:
Testing for Wait in FacilityHomepage

Rishith25 committed:
Merge branch 'develop' into cypress-tests-replace-cy.wait-with-cy.intercept

Rishith25 committed:
AssetHomepage tests

Rishith25 committed:
Add Precription intercept added

Rishith25 committed:
Wait added before medicine prescribed

Rishith25 committed:
Get request in consultationcreation

Rishith25 committed:
Logupdate

Rishith25 committed:
Merge branch 'develop' into cypress-tests-replace-cy.wait-with-cy.intercept

Rishith25 committed:
AssetHomePage Wait removal

Rishith25 committed:
FacilityHomepage.cy.ts test case wait

Rishith25 committed:
Patient Registration Waits

Rishith25 committed:
revert PatientInsurance.ts

Rishith25 committed:
Patient Consultation wait corrected

Rishith25 committed:
Paatient Discharge

Rishith25 committed:
Patient Consultation creation updated cypress

Rishith25 committed:
Patient discharge updated cypress

Rishith25 committed:
Edit Consultation cypress test intercept added

Rishith25 committed:
FacilityHomepage wait

Rishith25 committed:
Merge branch 'develop' into cypress-tests-replace-cy.wait-with-cy.intercept

Rishith25 committed:
Merge branch 'develop' into cypress-tests-replace-cy.wait-with-cy.intercept

Rishith25 committed:
PatientLogUpdate waits

Rishith25 committed:
PatientLogUpdate waits

Rishith25 committed:
PatientLogUpdate waits

Rishith25 committed:
PatientLogUpdate waits

Rishith25 committed:
Merge branch 'develop' into cypress-tests-replace-cy.wait-with-cy.intercept

Rishith25 committed:
PatientLogUpdate new Progress log update

Rishith25 committed:
PatientLogUpdate new Progress log update

Rishith25 committed:
PatientLogUpdate new Progress log update

Rishith25 committed:
PatientLogUpdate new Normal log update

Rishith25 committed:
PatientLogUpdate new Normal log update

Rishith25 committed:
PatientLogUpdate new Normal log update

Rishith25 committed:
Merge branch 'develop' into cypress-tests-replace-cy.wait-with-cy.intercept

Rishith25 committed:
PatientLogUpdate new Normal log update

Rishith25 committed:
PatientLogUpdate new Progress

Rishith25 committed:
Patient Registration cy.wait with intercepts

Rishith25 committed:
Patient Registration cy.wait with intercepts

Rishith25 committed:
Patient Discharge cy.wait

Rishith25 committed:
Patient LogUpdate reflection cards

Rishith25 committed:
Facility Homepage cy.wait

Rishith25 committed:
Facility Homepage cy.wait

Rishith25 committed:
Patient Doctor Notes, File Upload cy.wait

Rishith25 committed:
Patient Registration cy.waits

Rishith25 committed:
Patient File Upload cy.wait

Rishith25 committed:
Patient File Upload cy.wait

Rishith25 committed:
Patient File Upload cy.wait

Rishith25 committed:
Patient Registration wait

Rishith25 committed:
Patient Registration Edit Button

Rishith25 committed:
Patient Edit details

Rishith25 committed:
Patient File Upload with waits

Rishith25 committed:
Patient File Upload with waits

Rishith25 commented:
Hey @nihal467 
I replaced all instances of cy.wait() with cy.intercept() and dynamic waits.

However, the following issues arise when certain cy.wait() statements are removed:

In FacilityHomepage.cy.ts, removing cy.wait(2000) at line 177 causes the test case to fail because the logged-in nurse cannot see the notify message on the noticeboard. Additionally, intercepts are already included, so there is no possibility of using additional intercept here.

In PatientRegistration.tsx, removing cy.wait(3000) at line 158 results in a test case failure, indicating that after selecting the gender as "Female," the "Is antenatal?" checkbox is not displayed. No intercept is triggered when the gender selection is clicked.

In patientFileUpload.ts, removing cy.wait(2000) at line 30 within the verifyUploadFilePresence function causes the "User-level Based Permission for File Modification" test case to fail, while other test cases pass. I attempted to include intercepts, but the issue persists at the same point.

Could you please look into this?

nihal467 reviewed (changes requested):
- make the intercept and response as function , so these can be reused again for example cy.intercept("GET", "/api/v1/items/**").as("getItems"); can be placed inside a function interceptManageInventoryItem and cy.wait("@getItems").its("response.statusCode").should("eq", 200); can be placed inside a function verifyManageInventoryItem

Similar do that for all intercept

nihal467 commented on cypress/e2e/patient_spec/PatientConsultationDischarge.cy.ts:
```diff
@@ -76,10 +76,8 @@ describe("Patient Discharge based on multiple reason", () => {
-    cy.wait(2000);
     cy.clickSubmitButton("Confirm Discharge");
     cy.clickSubmitButton("Acknowledge & Submit");
-    cy.wait(2000);
```
- since you are removing all wait, add a intercept to verify the discharge is successful

Rishith25 committed:
Merge branch 'develop' into cypress-tests-replace-cy.wait-with-cy.intercept

Rishith25 committed:
Reusable functions for intercepts in cypress testcases

Rishith25 committed:
Reusable functions for intercepts in cypress testcases

Rishith25 committed:
User Manage Test Cases updated

Rishith25 committed:
User Manage Test Cases updated

Rishith25 committed:
Merge branch 'develop' into cypress-tests-replace-cy.wait-with-cy.intercept

Rishith25 committed:
UserManage test cases

Rishith25 committed:
UserManage

Rishith25 committed:
Merge branch 'develop' into cypress-tests-replace-cy.wait-with-cy.intercept

Rishith25 committed:
AddSkillButton with username

Rishith25 committed:
Merge branch 'develop' into cypress-tests-replace-cy.wait-with-cy.intercept

Rishith25 committed:
AssetsManage and AssetHomepage cypress tests

Rishith25 committed:
Merge branch 'develop' into cypress-tests-replace-cy.wait-with-cy.intercept

Rishith25 committed:
FacilityCreation cypress tests

Rishith25 committed:
Merge branch 'develop' into cypress-tests-replace-cy.wait-with-cy.intercept

Rishith25 committed:
Facility deletion test case resolve

Rishith25 committed:
Facility creation error resolved

Rishith25 committed:
fixed the cypress test

Rishith25 committed:
Merge branch 'develop' into cypress-tests-replace-cy.wait-with-cy.intercept

Rishith25 committed:
fixed the cypress test

Rishith25 committed:
fcility creation test

Rishith25 committed:
fcility creation test

Rishith25 committed:
Merge branch 'develop' into cypress-tests-replace-cy.wait-with-cy.intercept

Rishith25 committed:
Facility creation display error

Rishith25 committed:
Facility creation display error

Rishith25 committed:
Facility creation display error

Rishith25 committed:
Patient Discharge recovered test case

Rishith25 committed:
Facility Homepage bed capacity test case

Rishith25 committed:
Facility Creation test cases

Rishith25 committed:
Facility Creation test cases

Rishith25 committed:
Facility Creation test cases

Rishith25 committed:
Merge branch 'develop' into cypress-tests-replace-cy.wait-with-cy.intercept

Rishith25 committed:
commented notification for check

Rishith25 committed:
Facility creation test cases

Rishith25 commented:
@nihal467 Can you please help me to get it solved.

https://github.com/ohcnetwork/care_fe/actions/runs/12354446070/job/34475813772?pr=8963

I did not change anything in the `FacilityCreation.cy.ts`.
The test cases are always failing even after making required changes.

nihal467 committed:
Merge branch 'develop' into cypress-tests-replace-cy.wait-with-cy.intercept

Mahendar0701 committed:
Search UI for discharge patients (#9320)

JavidSumra committed:
Fixes UI Bug Issue in Symptom Editor (#9270)

Rishith25 committed:
Occupancy bed tooltip fully visible in mobile view (#9235)

Rishith25 committed:
Disallowing out of date range (#9201)

dependabot[bot] committed:
Bump @sentry/browser from 8.42.0 to 8.45.1 (#9473)

Bumps [@sentry/browser](https://github.com/getsentry/sentry-javascript) from 8.42.0 to 8.45.1.
- [Release notes](https://github.com/getsentry/sentry-javascript/releases)
- [Changelog](https://github.com/getsentry/sentry-javascript/blob/8.45.1/CHANGELOG.md)
- [Commits](https://github.com/getsentry/sentry-javascript/compare/8.42.0...8.45.1)

---
updated-dependencies:
- dependency-name: "@sentry/browser"
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

Rishith25 committed:
Patient registration form url intercept

Rishith25 committed:
Patient registration form url intercept

nihal467 committed:
Merge branch 'develop' into cypress-tests-replace-cy.wait-with-cy.intercept

nihal467 committed:
debug the facility creation

nihal467 committed:
removed hardcorded pagination verification

nihal467 commented on cypress/pageobject/utils/advanceFilterHelpers.ts:
```diff
@@ -4,17 +4,25 @@ export const advanceFilters = {
   },
 
   selectState(state: string) {
+    cy.wait(200);
```
dynamic handling is not possible currently, so for now its ok

nihal467 committed:
Merge branch 'develop' into cypress-tests-replace-cy.wait-with-cy.intercept

Rishith25 commented:
Thank you @nihal467 for helping me out.
